### PR TITLE
Animation compositions

### PR DESCRIPTION
**Pull request by dvoytenko**
A relatively small refactoring to allow animations to "include" other animations. The format extensions are described in the modified `amp-animation.md`.

Partial for #9129.

### CODE REVIEW

File: extensions/amp-animation/0.1/web-animations.js [hunk 1]
```diff
@@ -315,35 +333,123 @@ class Scanner {
...
-   * @param {boolean} validate
+   * @param {!Builder} builder
+   * @param {!CssContextImpl} css
+   * @param {!Array<string>} path
+   * @param {?Element} target
+   * @param {?Object<string, *>} vars
+   * @param {?WebAnimationTimingDef} timing
    */
-  constructor(win, rootNode, baseUrl, validate) {
+  constructor(builder, css, path, target, vars, timing) {
```
Comment (alanorozco) [inline extensions/amp-animation/0.1/web-animations.js]: `opt_timing`

File: extensions/amp-animation/0.1/web-animations.js [hunk 2]
```diff
@@ -355,35 +461,22 @@ export class MeasureScanner extends Scanner {
...
-   */
-  createRunner(resources) {
-    if (getMode().localDev || getMode().development) {
-      user().fine(TAG, 'Animation: ', this.requests_);
-    }
-    return this.unblockElements_(resources).then(() => {
-      return new WebAnimationRunner(this.requests_);
-    });
+    /** @const @private {!Array<!Promise>} */
+    this.deps_ = [];
```
Comment (alanorozco) [inline extensions/amp-animation/0.1/web-animations.js]: deps for what?

File: extensions/amp-animation/0.1/web-animations.js [hunk 3]
```diff
@@ -502,31 +624,46 @@ export class MeasureScanner extends Scanner {
...
-    this.timing_ = timing;
-    callback();
+
+    // Push new context and perform calculations.
+    const targets =
+        (spec.target || spec.selector) ?
+        this.resolveTargets_(spec) :
+        [null];
+    targets.forEach(target => {
+      this.target_ = target || prevTarget;
```
Comment (alanorozco) [inline extensions/amp-animation/0.1/web-animations.js]: why does this fallback to `prevTarget`?

File: extensions/amp-animation/amp-animation.md [hunk 4]
```diff
@@ -299,6 +333,64 @@ can be reduced to an array of components. For instance:
...
+### Animation composition
+
+Animations can reference other animations thus combining several `amp-animation` declarations into a single final animation.
+
+For instance:
+```html
+<amp-animation id="anim1" layout="nodisplay">
+<script type="application/json">
+{
+  "animamtion": "anim2",
```
Comment (alanorozco) [inline extensions/amp-animation/amp-animation.md]: typo: `animation`

File: extensions/amp-animation/amp-animation.md [hunk 5]
```diff
@@ -299,6 +333,64 @@ can be reduced to an array of components. For instance:
...
+without a target (`selector`). In such case, the included animation is expected to reference its own target.
+
+Another form allows the including animation to provide the target or multiple targets. In that case, the included
+animation is executed for each matched target. For instance:
+```html
+<amp-animation id="anim1" layout="nodisplay">
+<script type="application/json">
+{
+  "selector": ".target-class",
+  "animamtion": "anim2",
```
Comment (alanorozco) [inline extensions/amp-animation/amp-animation.md]: ditto

File: extensions/amp-animation/amp-animation.md [hunk 6]
```diff
@@ -299,6 +333,64 @@ can be reduced to an array of components. For instance:
 </amp-animation>
 ```
 
+
+### Animation composition
```
Comment (alanorozco) [inline extensions/amp-animation/amp-animation.md]: Can this clarify further on the difference between setting `animation` as a reference vs. `animations`? Some confusion can come from the fact that they can be described top-level or nested one level down.

File: extensions/amp-animation/0.1/web-animations.js [hunk 7]
```diff
@@ -315,35 +333,123 @@ class Scanner {
...
+   * @param {string} baseUrl
+   * @param {!../../../src/service/vsync-impl.Vsync} vsync
+   * @param {!../../../src/service/resources-impl.Resources} resources
+   */
+  constructor(win, rootNode, baseUrl, vsync, resources) {
+    /** @const @private */
+    this.css_ = new CssContextImpl(win, rootNode, baseUrl);
+
+    /** @const @private */
+    this.vsync_ = vsync;
```
Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: Easier to test. This is an internal class - the main reason is to split the code and test.

File: extensions/amp-animation/0.1/web-animations.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: It's not optional. Only nullable.

File: extensions/amp-animation/0.1/web-animations.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: Dependencies to fully resolve all animation requests. Some can only be done asynchronously. I added the description.

File: extensions/amp-animation/0.1/web-animations.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: `prevTarget` is basically the "parent" target. Either a nested animation overrides it or the "parent" target is used. The `mergeVars_` and `mergeTiming_` that follow, essentially do the same thing.

File: extensions/amp-animation/amp-animation.md (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-animation/amp-animation.md]: Added possible reasons. Similar to function call composition, two reasons: allow reuse across multiple animations (as with calling function from different places) or simply making each animation more manageable (as with splitting unwieldy function into two, even if there's no reuse).

File: extensions/amp-animation/amp-animation.md (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-animation/amp-animation.md]: Done

File: extensions/amp-animation/amp-animation.md (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-animation/amp-animation.md]: Done

**Comment by dvoytenko** [comment]
@alanorozco All comments addressed. PTAL.

**Pull request merged**